File: src/main/java/com/hiddewieringa/elevator/domain/elevator/model/ElevatorAggregate.java
package com.hiddewieringa.elevator.domain.elevator.model;

import com.hiddewieringa.elevator.domain.elevator.command.CallElevator;
import com.hiddewieringa.elevator.domain.elevator.command.CreateElevator;
import com.hiddewieringa.elevator.domain.elevator.event.ElevatorCalled;
import com.hiddewieringa.elevator.domain.elevator.event.ElevatorCreated;
import com.hiddewieringa.elevator.domain.elevator.query.FloorResult;
import com.hiddewieringa.elevator.domain.elevator.query.QueryFloor;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.commandhandling.model.AggregateIdentifier;
import org.axonframework.eventhandling.EventHandler;
import org.axonframework.queryhandling.QueryHandler;
import org.axonframework.queryhandling.QueryMessage;
import org.axonframework.spring.stereotype.Aggregate;

import static org.axonframework.commandhandling.model.AggregateLifecycle.apply;

@Aggregate
public class ElevatorAggregate {

    @AggregateIdentifier
    private ElevatorId id;

    private long floor = 0L;

    private ElevatorAggregate() {
    }

    @CommandHandler
    public ElevatorAggregate(CreateElevator command) {
        apply(new ElevatorCreated(command.getElevatorId()));
    }

    @EventHandler
    public void on(ElevatorCreated event) {
        this.id = event.getElevatorId();
    }

    @CommandHandler
    public void call(CallElevator command) {
        apply(new ElevatorCalled(command.getElevatorId(), command.getFloor()));
    }

    @EventHandler
    public void elevatorCalled(ElevatorCalled event) {
        this.floor = event.getFloor();
    }

    @QueryHandler(queryName = "test")
    public FloorResult queryFloor(QueryMessage<QueryFloor, FloorResult> queryMessage) {
//        org.axonframework.commandhandling.model.Aggregate<ElevatorAggregate> q= elevatorAggregateRepository.load(query.getElevatorId().toString());
        return new FloorResult(floor);
    }

    @QueryHandler(queryName = "test")
    public FloorResult queryFloor2(QueryFloor queryFloor) {
//        org.axonframework.commandhandling.model.Aggregate<ElevatorAggregate> q= elevatorAggregateRepository.load(query.getElevatorId().toString());
        return new FloorResult(floor);
    }

    public long getFloor() {
        return floor;
    }
}

File: src/main/java/com/hiddewieringa/elevator/domain/elevator/model/ElevatorAggregate.kt
package com.hiddewieringa.elevator.domain.elevator.model

import com.hiddewieringa.elevator.domain.elevator.*
import org.axonframework.commandhandling.CommandHandler
import org.axonframework.commandhandling.model.AggregateIdentifier
import org.axonframework.commandhandling.model.AggregateLifecycle.apply
import org.axonframework.eventhandling.EventHandler
import org.axonframework.spring.stereotype.Aggregate

@Aggregate
class ElevatorAggregate {

    @AggregateIdentifier
    private var id: ElevatorId? = null

    private var floor = 0L

    private var doorsOpened = false

    private constructor()

    @CommandHandler
    constructor(command: CreateElevator) {
        apply(ElevatorCreated(command.elevatorId))
    }

    @EventHandler
    fun handle(event: ElevatorCreated) {
        this.id = event.elevatorId
    }

    @CommandHandler
    fun handle(command: OpenDoors) {
        if (doorsOpened) {
            throw IllegalArgumentException("Doors already open")
        }
        apply(ElevatorDoorsOpened(command.elevatorId))
    }

    @EventHandler
    fun handle(event: ElevatorDoorsOpened) {
        doorsOpened = true
    }

    @CommandHandler
    fun handle(command: CloseDoors) {
        if (!doorsOpened) {
            throw IllegalArgumentException("Doors already closed")
        }
        apply(ElevatorDoorsClosed(command.elevatorId))
    }

    @EventHandler
    fun handle(event: ElevatorDoorsClosed) {
        doorsOpened = false
    }

    @CommandHandler
    fun handle(command: CallElevator) {
        if (floor == command.floor) {
            // Floor is same as calling floor
            apply(ElevatorMovedToFloor(command.elevatorId, floor))
            return;
        }
        while (floor != command.floor) {
            apply(ElevatorMovedToFloor(command.elevatorId, if (floor > command.floor) floor - 1 else floor + 1))
        }
    }

    @EventHandler
    fun handle(event: ElevatorMovedToFloor) {
        floor = event.floor
    }
}
